File: Bloomier/bloomierFilter.py
#from core.bloomierHasher import *
#import bloomierHasher
from core.bloomierHasher import BloomierHasher
from core.orderAndMatch import *
from core.orderAndMatchFinder import *
from core.utilEncode import *
from core.util import *

class BloomierFilter:
    def __init__(self, hashSeed, keysDict, m, k, q, useTable = False):
        self.hashSeed = hashSeed
        self.keysDict = keysDict
        self.m = m
        self.k = k
        self.q = q
        self.hasher = BloomierHasher(hashSeed, m, k, q)
        self.byteSize = getByteSize(q)
        
        oamf = OrderAndMatchFinder(hashSeed, keysDict, m, k, q)
        oam =  oamf.find()
        
        ## TODO - this can be just bytearray 
        self.table = [[0] * self.byteSize] * m
        self.valueTable = [0] * m

        if useTable is False:
            self.create(keysDict, oam)
        
    def getTable(self): return self.table
    def setTable(self, table): self.table = table
    def getValueTable(self): return self.valueTable
    def setValueTable(self, table): self.valueTable = table
        
    def xorOperations(self, value, M, neighbors):
        #value = [0] * self.byteSize
        #print "???", value
        byteArrayXor(value, M)
        
        for v in neighbors:
            byteArrayXor(value, self.table[v])
        
        return value
        
    def get(self, key):
        neighbors = self.hasher.getNeighborhood(key)
        mask = self.hasher.getM(key)
        
        #print neighbors
        valueToGet = [0] * self.byteSize
        self.xorOperations(valueToGet, mask, neighbors)
        
        h = decode(valueToGet) # , self.byteSize)
        
        try:
            L = neighbors[h]
            return self.valueTable[L]
        except IndexError:
            return None
            
    def set(self, key, value):
        neighbors = self.hasher.getNeighborhood(key)
        mask = self.hasher.getM(key)
        
        #print neighbors
        valueToGet = [0] * self.byteSize
        self.xorOperations(valueToGet, mask, neighbors)
        
        h = decode(valueToGet) # , self.byteSize)
        
        try:
            L = neighbors[h]
            self.valueTable[L] = value
            return True
        except IndexError:
            return False    
            
    def create(self, map, oam):
        assert (oam is not None)
        piList = oam.piList
        tauList = oam.tauList
        #print pi, tau
        
        for i, pi in enumerate(piList):
        #for pi in piList:
            key = pi
            value = map[key]
            #print value
            neighbors = self.hasher.getNeighborhood(key)
            mask = self.hasher.getM(key)
            l = tauList[i] # tauList contains the iota values
            L = neighbors[l]
            
            encodeValue = encode(l, self.byteSize)
            valueToStore = [0] * self.byteSize
            
            byteArrayXor(valueToStore, encodeValue)
            byteArrayXor(valueToStore, mask)
            
            for i, v in enumerate(neighbors):
                if i == l:
                    pass
                else:
                    # The h_value in the table should be applied
                    byteArrayXor(valueToStore, self.table[v])

            self.table[L] = valueToStore
            self.valueTable[L] = value
        #print self.valueTable
        
    def table2string(self):
        """
        transforms table into byte array
        """
        #self.table = [[0] * self.byteSize] * m
        #self.valueTable = [0] * m
        result = chr(self.byteSize) + int2string(self.m) + chr(4) # header are 6 bytes (5 bytes + 1 for the valueTable width)
        #                                                      ^ <-- Wwe use 4 byte for value storage
        for row in self.table:
            #print row
            temp = byteArray2string(row)
            #sprint(temp)
            result += temp

        #print len(result)
        #result = int2string(len(result)) + result
        for row in self.valueTable:
            #print(row)
            result += int2string(row, 4) # or int2string(row, 2) to store 2 bytes only

        return result
        
    def string2table(self, string):
        assert type(string) is str
        # 9 bytes of header should be analyzed
        startOfData = 6
        byteSize = ord(string[0])
        m = string2int(string[1:5])
        valueSize = string2int(string[5:startOfData])

        lengthOfFirstTable = m * byteSize

        firstTable = string[startOfData:startOfData + lengthOfFirstTable]
        valueTable = string[startOfData + lengthOfFirstTable:]

        t1 = []
        t2 = []
        for i in range(m):
            # firstTable[i*byteSize:(i+1)*byteSize] returns array of chars ['\x01'...]
            temp1 = [ord(x) for x in firstTable[i*byteSize:(i+1)*byteSize]]
            t1.append(temp1)
            #print len(valueTable[i*valueSize:(i+1)*valueSize])
            temp2 = string2int(valueTable[i*valueSize:(i+1)*valueSize])
            t2.append(temp2)
        # setup the table now
        return t1, t2
        
if __name__ == "__main__":
    sys.path.append("../test")
    from testBloomierFilter import *
    
    unittest.main(verbosity=2)

File: ScanningEngine/Cpp/libs/bloomierFilter/BloomierFilter.cpp
//
// Created by smcho on 5/30/13.
// Copyright (c) 2013 ___MPC___. All rights reserved.
//
// To change the template use AppCode | Preferences | File Templates.
//


#include "BloomierFilter.h"

namespace bloomier {

    BloomierFilter::BloomierFilter(int hashSeed, std::map<std::string, int>* keyMap, int m, int k, int q)
    {
        this->hashSeed = hashSeed;
        this->keyMap = keyMap;
        this->m = m;
        this->k = k;
        this->q = q;

        //std::cout << "K" << this->k << endl;
        byteSize = Util::getByteSize(q);
        oam = new OrderAndMatchFinder(hashSeed, keyMap, m, k, q);
        om = oam->find();
        h = new BloomierHasher(hashSeed, m, k, q);

        table = new unsigned char[byteSize * m];
        // not size optimal, only for testing purposes
        valueTable = new int[m];

        create(keyMap, oam);
    }

    BloomierFilter::~BloomierFilter()
    {
        if (oam != NULL) delete oam;
        // if (om != NULL) delete om; om is deleted automatically from oam.
        if (table != NULL) delete table;
        if (valueTable != NULL) delete valueTable;
        if (h != NULL) delete h;
    }

    void BloomierFilter::create(std::map<std::string, int>* keyMap, OrderAndMatchFinder* oam)
    {
        auto piList = oam->getPiList();
        auto tauList = oam->getTauList();

        unsigned char neighbors[k];
        unsigned char mask[byteSize];
        unsigned char encodedValue[byteSize];

        //Util::print(piList);
        //Util::print(tauList);

        //int i = 0;
        int keySize = piList->size();
        for (int i = 0; i < keySize; i++)
            // auto ip = piList->begin(); ip != piList->end(); ++ip)
        {
            std::string key = (*piList)[i];
            h->getNeighborhood(key, neighbors);
            //Util::print(neighborhood, k);
            h->getM(key, mask, byteSize);
            //Util::print(mask, byteSize);

            int value = (*keyMap)[key];
            int l = (*tauList)[i];
            int L = neighbors[l];

            UtilEncode::encode(l, byteSize, encodedValue);
            //Util::print(encodedValue, byteSize);

            Util::byteArrayXor(encodedValue, mask, byteSize);

            for (int i = 0; i < k; i++)
            {
                if (i == l) continue;
                int v = neighbors[i];
                Util::byteArrayXor(encodedValue, table + byteSize*v, byteSize);
            }

            Util::setInArray(table + byteSize*L, encodedValue, byteSize);
            valueTable[L] = value;
            //int index =
            //std::cout << i << "?" << key << ':' << value << "*" << l << std::endl;
            //i++;
        }
    }

/*
 * Code duplication with get value
 */
    bool BloomierFilter::set(std::string key, int value)
    {
        unsigned char neighbors[k];
        unsigned char mask[byteSize];
        unsigned char valueToGet[byteSize]; // = {0};
        memset(valueToGet, 0, sizeof(unsigned char)*byteSize);

        h->getNeighborhood(key, neighbors);
        h->getM(key, mask, byteSize);

        Util::byteArrayXor(valueToGet, mask, byteSize);
        for (int i = 0; i < k; i++)
        {
            int v = neighbors[i];
            Util::byteArrayXor(valueToGet, table + byteSize*v, byteSize);
        }

        //Util::xorOperations(valueToGet, mask, n)
        int l = UtilEncode::decode(valueToGet, byteSize);

        if (l < m) {
            int L = neighbors[l];
            valueTable[L] = value;
            return true;
        }

        return false;
    }

    bool BloomierFilter::get(std::string key, int& value)
    {
        unsigned char neighbors[k];
        unsigned char mask[byteSize];
        unsigned char valueToGet[byteSize]; // = {0};
        memset(valueToGet, 0, sizeof(unsigned char)*byteSize);

        h->getNeighborhood(key, neighbors);
        h->getM(key, mask, byteSize);

        Util::byteArrayXor(valueToGet, mask, byteSize);
        for (int i = 0; i < k; i++)
        {
            int v = neighbors[i];
            Util::byteArrayXor(valueToGet, table + byteSize*v, byteSize);
        }

        //Util::xorOperations(valueToGet, mask, n)
        int l = UtilEncode::decode(valueToGet, byteSize);

        if (l < m) {
            int L = neighbors[l];
            value = valueTable[L];
            return true;
        }

        return false;
    }

}